docs/proposals/cluster-mesh.md

```diff
@@ -0,0 +1,101 @@
+
+- High availability of clusters (failover of a service to a different cluster, possibly in a different datacenter),
+- Easy workload migration between clusters,
+- Hybrid cloud scenarios, like providing special services by on-premises clusters to cloud clusters, or vice-versa.
+
+This should be in the ideal case possible without the need for too complex setup on the cluster level and service level (e.g. with automatic service discovery), and in a secure manner (e.g. network policies should work across multiple clusters, traffic between the clusters can be encrypted).
+
+All of this can be achieved by relying on the feature of the Cilium CNI called Cluster Mesh.
+
+## Implementation
+
+We will add a new CRD on the Seed level (`ClusterMesh`), that would define a list of clusters that are part of the mesh.
+
+In the initial implementation it will be possible to mesh only clusters that belong to the same KKP seed cluster, to avoid flow of data from the user clusters / KKP seed clusters towards the KKP main cluster. This can be enhanced without major changes in the future, e.g. if Cilium Cluster Mesh deployment will be simplified.
+
+The `ClusterMesh` CRD will be defined as follows:
+
+```go
+// ClusterMesh is the object representing a mesh of multiple KKP user clusters.
+type ClusterMesh struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  Spec ClusterMeshSpec `json:"spec,omitempty"`
+}
+
+// ClusterMeshSpec specifies the mesh of KKP user clusters.
+type ClusterMeshSpec struct {
+  // Clusters is a list of clusters connected in the ClusterMesh.
+  Clusters []ClusterMeshCluster `json:"clusters,omitempty"`
+}
+
+// ClusterMeshCluster contains mesh configuration for a cluster.
+type ClusterMeshCluster struct {
+  // ID an internal identifier of the cluster in the ClusterMesh.
+  // Must be unique across clusters in a ClusterMesh.
+  ID int `json:"ID,omitempty"`
+
+  // APIServerServiceType defines the k8s service type used to expose the clustermesh apiserver the other clusters.
+  APIServerServiceType v1.ServiceType `json:"apiServerServiceType,omitempty"`
+
+  // APIServerIPs contains a list of IPs on which the clustermesh apiserver of this cluster is exposed to the other clusters.
+  // It can be a single IP if APIServerServiceType == LoadBalancer, or multiple IPs if APIServerServiceType == NodePort.
+  APIServerIPs []string `json:"apiServerIPs,omitempty"`
+
+  // APIServerPort defines the port on which the clustermesh apiserver of this cluster is exposed to the other clusters.
+  APIServerPort int `json:"apiServerPort,omitempty"`
+}
+```
+
+A new cluster-mesh controller on the seed level  (EE only) will make sure that:
+
+- For each `ClusterMesh`, cluster mesh CA certificates are issued and stored in a secret in the kubermatic namespace,
+- Each cluster in the `ClusterMesh` list is assigned with an unique cluster ID (1 byte) and stored in the `ClusterMesh` CR,
```

Comment:
I don't think the `cluster-mesh-controller` should alter the `ClusterMesh` spec. Maybe we should have a status field for information the controller produces and needs to remember?

Reply:
Fair point. I originally thought abut having the ability to optionally define some stuff like clusterMesh ID by the user, but its probably to make an exact distinction on what needs to be defined by the user / vs assigned by controller (via status). Will update that.